RUN: enable --show-output when executing tests

emmericp opened the pull request:
Fixes https://github.com/intellij-rust/intellij-rust/issues/3994

This requires this patch for rust (hopefully in 1.38?): https://github.com/rust-lang/rust/pull/62600

This is work in progress because it doesn't work yet, I'm looking for some help here:
The problem is you can't run `rustVersion()` in the UI thread and I'd really like to have this behavior as default if the compiler is new enough.
What's the best way to achieve this? Fallback would be to make this an option for the run configuration, but that would be a little bit annoying since virtually everyone would want to enable this?


Edit: changed the code to hardcode 1.38 as Rust version passed to `patchArgs` because I'm using this branch ;) Looking for a good way to get the Rust version here.

<!--
Hello and thank you for the pull request!

We don't have any strict rules about pull requests, but you might check
https://github.com/intellij-rust/intellij-rust/blob/master/CONTRIBUTING.md
for some hints!

Note that we need an electronic CLA for contributions:
https://github.com/intellij-rust/intellij-rust/blob/master/CONTRIBUTING.md#cla

After you sign the CLA, please add your name to
https://github.com/intellij-rust/intellij-rust/blob/master/CONTRIBUTORS.txt

:)
-->

mchernyavsky commented on src/main/kotlin/org/rust/cargo/runconfig/CargoTestRunState.kt:
```diff
@@ -75,6 +76,9 @@ class CargoTestRunState(
             } else {
                 libtestArgs.add(noFormatJson)
             }
+            if (rustcVer != null && rustcVer >= SemVer("1.38.0", 1, 38, 0) && !libtestArgs.contains("--show-output")) {
```
Maybe we also need to check current channel?

emmericp commented:
The feature has been merged and is available in Rust 1.39 nightly builds with a build date of 28-08-2019 or later. I've updated this PR accordingly and fixed the merge conflict

Undin commented on src/main/kotlin/org/rust/cargo/runconfig/CargoTestRunState.kt:
```diff
@@ -60,7 +66,24 @@ class CargoTestRunState(
+
+        private fun checkShowOutputSupport(ver: RustcVersion?): Boolean {
+            if (ver == null) return false
+            // --show-output is supported since 1.39.0-nightly/dev with a build date later than 2019-08-27
```
Can we check that this flag is supported without version checking? Like we do with some other flags

mchernyavsky replied:
I think no. The `--show-output` is only mentioned in `cargo test -- --help` output, but this command may trigger compilation.

emmericp replied:
i don't think it's possible to detect this without compiling tests as it's a feature in libtest that's only in the resulting test binary

mchernyavsky commented:
bors r+
